app/expirynotify_test.go

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package app
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/stretchr/testify/require"
+)
+
+func TestNotifySessionsExpired(t *testing.T) {
+	th := Setup(t).InitBasic()
+	defer th.TearDown()
+
+	handler := &testPushNotificationHandler{t: t}
+	pushServer := httptest.NewServer(
+		http.HandlerFunc(handler.handleReq),
+	)
+	defer pushServer.Close()
+
+	th.App.UpdateConfig(func(cfg *model.Config) {
+		*cfg.EmailSettings.PushNotificationServer = pushServer.URL
+	})
+
+	t.Run("push notifications disabled", func(t *testing.T) {
+		th.App.UpdateConfig(func(cfg *model.Config) {
+			*cfg.EmailSettings.SendPushNotifications = false
+		})
+
+		err := th.App.NotifySessionsExpired()
+		// no error, but also no requests sent
+		require.Nil(t, err)
+		require.Equal(t, 0, handler.numReqs())
+	})
+
+	t.Run("two sessions expired", func(t *testing.T) {
+		th.App.UpdateConfig(func(cfg *model.Config) {
+			*cfg.EmailSettings.SendPushNotifications = true
+		})
+
+		data := []struct {
+			deviceId  string
+			expiresAt int64
+			notified  bool
+		}{
+			{deviceId: "android:11111", expiresAt: model.GetMillis() + 100000, notified: false},
+			{deviceId: "android:22222", expiresAt: model.GetMillis() - 1000, notified: false},
+			{deviceId: "android:33333", expiresAt: model.GetMillis() - 2000, notified: false},
+			{deviceId: "android:44444", expiresAt: model.GetMillis() - 3000, notified: true},
+		}
+
+		for _, d := range data {
+			_, err := th.App.CreateSession(&model.Session{
+				UserId:        th.BasicUser.Id,
+				DeviceId:      d.deviceId,
+				ExpiresAt:     d.expiresAt,
+				ExpiredNotify: d.notified,
+			})
+			require.Nil(t, err)
+		}
+
+		err := th.App.NotifySessionsExpired()
+
+		require.Nil(t, err)
+		require.Equal(t, 2, handler.numReqs())
+
+		expected := []string{"22222", "33333"}
+		require.Equal(t, model.PUSH_TYPE_SESSION, handler.notifications()[0].Type)
+		require.Contains(t, expected, handler.notifications()[0].DeviceId)
+		require.Contains(t, handler.notifications()[0].Message, "Session Expired")
+
+		require.Equal(t, model.PUSH_TYPE_SESSION, handler.notifications()[1].Type)
+		require.Contains(t, expected, handler.notifications()[1].DeviceId)
+		require.Contains(t, handler.notifications()[1].Message, "Session Expired")
```

Comment:
I think we should have another test that checks for errors

Reply:
Thanks @ethervoid .  The only error that can come out of `App.NotifySessionExpired` is if the SessionStore fails on a basic select.  I would have to mock the SessionStore in a way that allows me to fail on that select, and then I would just be checking that `App.NotifySessionExpired` returns the error.  I think that would be a lot of work for not much validation... unless you have an idea for a test that adds value.

Reply:
Sorry, I didn't write it very well. I meant to test the `NotificationsLogError()` part [here](https://github.com/mattermost/mattermost-server/pull/14732/files#diff-7ba161c09bdeff2e097acaee45fcc60fR46) so we can validate using tests that the error logs is working correctly and the flow for multiple notifications works properly even though we have an error in one of them

Reply:
`a.NotificationsLog()` is just a mlog.Logger instance.  It's used in a bunch of places other places as well.  I guess I'm not getting what this test would entail and what the value is.  Are you saying I should test that the mlog.Logger logged the right stuff?  I don't that done anywhere else we do logging.  What would the value be?